File: python/webhook.py
from flask import Flask, request, jsonify, make_response
from Crypto.PublicKey import RSA
from Crypto.Signature import PKCS1_v1_5
from Crypto.Hash import SHA256
from base64 import b64decode

from python.signature import send_signed_request

app = Flask(__name__)


@app.route('/')
def start():
    return 'Webhooks with Python'


@app.route('/notify', methods=['POST'])
def webhook_notification():
    print('Received Webhook Notification!')
    verify_signature(request.headers, request.data)
    data = {'message': 'Done', 'code': 'SUCCESS'}
    return make_response(jsonify(data), 200)


def get_pub_key():
    response = send_signed_request(
        'POST',
        '/binancepay/openapi/certificates',
    )
    return response['data'][0]['certPublic']


def verify_signature(headers, data):
    print('Verifying Signature...')
    print(headers)
    print(data)
    timestamp, nonce, signature = headers['Binancepay-Timestamp'], headers['Binancepay-Nonce'], headers[
        'Binancepay-Signature']
    payload = timestamp + '\n' + nonce + '\n' + data.decode('utf-8') + '\n'
    pub_key = get_pub_key()
    rsakey = RSA.importKey(pub_key)
    signer = PKCS1_v1_5.new(rsakey)
    digest = SHA256.new()
    digest.update(payload.encode('utf-8'))
    if signer.verify(digest, b64decode(signature)):
        print('Signature Verified!')
        return
    print("Signature Not Verified!")
    return


if __name__ == '__main__':
    app.run(debug=True)

File: node+ts+express/src/webhook.ts
import axios from "axios";
import crypto from "crypto";
import hmac from "crypto-js/hmac-sha512";
import { Application } from "express";

const randomCharts = (length: number): string => {
  let result = "";
  const characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
};

const generateSignedPayload = (
  payload: string
): { nonce: string; timestamp: number; signature: string } => {
  const timestamp = Date.now();
  const nonce = randomCharts(32);
  const payload_to_sign = `${timestamp}\n${nonce}\n${payload}\n`;
  const signature = hmac(
    payload_to_sign,
    process.env.BINANCE_SECRET_API_KEY || ""
  )
    .toString()
    .toUpperCase();

  return { nonce, timestamp, signature };
};

const verifyWebhookSignature = async (
  binancePayTimestamp: string,
  binancePayNonce: string,
  bianacePaySignature: string,
  data: any
): Promise<boolean> => {
  console.log("Verifying Signature...");
  const payload = `${binancePayTimestamp}\n${binancePayNonce}\n${JSON.stringify(
    data
  )}`;
  const publicKey = await getPubKey();
  console.log("PublicKey: ", publicKey);

  const decodedSignature = Buffer.from(bianacePaySignature, "base64");

  const verify = crypto.createVerify("SHA256");
  verify.write(payload);
  verify.end();

  return await verify.verify(publicKey, decodedSignature);
};

const getPubKey = async (): Promise<string> => {
  const { nonce, timestamp, signature } = generateSignedPayload(
    JSON.stringify({})
  );
  const result = await axios.post(
    "https://bpay.binanceapi.com/binancepay/openapi/certificates",
    {},
    {
      headers: {
        "BinancePay-Certificate-SN": process.env.BINANCE_API_KEY,
        "Content-Type": "application/json;charset=utf-8",
        "BinancePay-Timestamp": timestamp,
        "BinancePay-Nonce": nonce,
        "BinancePay-Signature": signature,
      },
    }
  );

  return result.data.data[0].certPublic;
};

export const loadWebHookEndpoints = (app: Application): void => {
  app.post("/binance/order/notification", async (req, res) => {
    const verified = await verifyWebhookSignature(
      req.headers["binancepay-timestamp"] as string,
      req.headers["binancepay-nonce"] as string,
      req.headers["binancepay-signature"] as string,
      req.body
    );

    console.log(
      "********************BINANCE WEBHOOK CALL*******************",
      req.body
    );
    if (verified) {
      return res.status(200).send({
        returnCode: "SUCCESS",
        returnMessage: null,
      });
    } else {
      return res
        .status(401)
        .send({ returnCode: "ERROR", returnMessage: "Not authorized" });
    }
  });
};
